lib/LTP/WhiteList.pm

```diff
@@ -97,4 +96,46 @@ sub is_test_disabled {
     return 0;
 }
 
+sub find_whitelist_testsuite {
+    my ($suite) = @_;
+
+    my $path = get_var('LTP_KNOWN_ISSUES');
+    return undef unless defined($path) and -e $path;
+
+    my $content = path($path)->slurp;
+    my $issues  = Mojo::JSON::decode_json($content);
+    return undef unless $issues;
+    return $issues->{$suite};
+}
+
+sub list_skipped_tests {
+    my ($env, $suite) = @_;
+    my @skipped_tests;
+    $suite = find_whitelist_testsuite($suite);
+    return @skipped_tests unless ($suite);
+
+    die 'Unsupported format for `list_skipped_tests()`' if (ref($suite) eq 'ARRAY');
+
+    for my $test (keys(%$suite)) {
+        my @entrys = grep { $_->{skip} && whitelist_entry_match($_, $env) } @{$suite->{$test}};
+        push @skipped_tests, $test if @entrys;
+    }
+    return @skipped_tests;
+}
+
+sub whitelist_entry_match
+{
+    my ($entry, $env) = @_;
+    my @mandatory_attributes = qw(product ltp_version revision arch kernel backend retval flavor);
+
+    if (grep { !exists $env->{$_} } @mandatory_attributes) {
```

Comment:
I'm not sure if we wanna be that strict here. The old version just produced a error like:
```
Use of uninitialized value $env->{$filter} in pattern match (m//) at -e line 1.
```
So maybe, we want just skip a filter, if the environment doesn't have that value?!
@baierjan  WDJT?

Reply:
Wow, this code changed a lot since I saw it last time... 😄  

Yes, it would be nice to skip missing filters to not generate warnings, however as far as I know, none of them is really a mandatory filter (and usually only a few are used for a single entry). The only mandatory attribute is the name of the test (variables `$suite` and `$test` from the `find_whitelist_entry`) no other assumption can be made. So making the attributes mandatory is not a good choice.

If you need more deep clarification, @mdoucha can help you here as he is more familiar with the code and current usage.

Reply:
Thx, I will put @mdoucha into the reviewer list. To make my point clear here is a example:

ltp_known_issues.json:
```json
{
  "syscalls": {
    "ioctl_loop01": [
      {
        "message": "Kernel bug. WONTFIX for livepatches. bsc#1170794",
        "retval": "^1$",
        "product": "sle:15-SP2$",
        "revision": "kernel-livepatch-SLE15-SP2_Update_0$"
      }
    ]
}
```

The given `$env` would look like:
```perl
{
   product => 'sle:15-SP2',
   revision => 'kernel-livepatch-SLE15-SP2_Update_666'
   retval => 1
}
```

When we now call `whitelist_entry_match()` on that entry, it won't match because the `revision` doesn't.

**The question is, how should `whitelist_entry_match()` acct, if the `revision` isn't inside the `$env` like:**
```perl
whitelist_entry_match(
     {
        message => "Kernel bug. WONTFIX for livepatches. bsc#1170794",
        retval => '^1$',
        product =>  'sle:15-SP2$',
        revision => 'kernel-livepatch-SLE15-SP2_Update_0$'
      },
      {
        product => 'sle:15-SP2',
      }
);
```

Reply:
While writing this, I think we should **not** `die()` if the env doesn't have all filters set, but plz comment!

Reply:
The whitelist entry should match if all provided filters match, so I would say if any of the filter is not in the environment, it is probably impossible for the filter to match (unless the filter says `'^$'`, but that's in my opinion a useless filter and can be omitted completely). Correct me if I am wrong, but the old version just emitted a warning and did the match with an empty string, I pretty much think that is the expected behavior. It would be probably nice to create a user-readable warning to let the user know, he is using either a non-standard filter or filter which is not valid in this context (and of course there might be also a typo in that filter).

We should probably emit `record_info` about missing env and assume an empty string for the filter evaluation, no reason to `die()`. Do you agree?

Reply:
Yes I agree that we should not `die()`.

The old implementation did the match with a `undef` string (so warning appeared), which is always false, or? 

So we would like to keep that behavior, that the above example should *not* match, because we miss `retval` and `revision` in the environment?!

Reply:
> The old implementation did the match with a `undef` string (so warning appeared), which is always false, or? 

Yes, I would think so.

> So we would like to keep that behavior, that the above example should *not* match, because we miss `retval` and `revision` in the environment?!

Exactly, we want to match for the specific environment only. If they are missing, it is different environment and we should *not* match. Again, @mdoucha can correct me, if my assumptions are wrong.

Reply:
The old implementation was [paired with code](https://github.com/os-autoinst/os-autoinst-distri-opensuse/blob/a910395f0804f8dc0b1c003f0c57afa9355e90a1/lib/LTP/utils.pm#L234) which always set all the environment variables to _something_. If you want to allow whitelist lookup with incomplete environment, which is *not* currently supported at all, the I suggest treating undefined env variables as silent match failure.

Reply:
Thanks for pointing it out and correcting me! I will change the code.